src/VisualStudio/Core/Def/Implementation/TableDataSource/TableItem.cs

```diff
@@ -74,41 +81,34 @@ public TableItem(IEnumerable<TableItem<T>> items)
             _cache = SharedInfoCache.GetOrAdd(collectionHash, orderedItems, c => new SharedInfoCache(c.Select(i => i.PrimaryDocumentId).ToImmutableArray()));
         }
 
+        // item must be either one of diagnostic data and todo item
         public DocumentId PrimaryDocumentId
-        {
-            get
-            {
-                return Extensions.GetDocumentId(Primary);
-            }
-        }
+            => (Primary is DiagnosticData diagnostic) ? diagnostic.DocumentId : ((TodoItem)(object)Primary).DocumentId;
```

Comment:
man do i dislike this.

Reply:
Me too, will follow up with more refactoring to clean this up.

Reply:
ya. you can make TableItem to hold onto document and projectId directly rather than get it through item or pass in lambda to get those or make diagnostic/todoItem to implement interface to get document/project id.

Reply:
Fixed.